src/lib/io/VDB.cpp

```diff
@@ -0,0 +1,371 @@
+    static struct OpenVDBLib {
+        OpenVDBLib()  { openvdb::initialize(); }
+        ~OpenVDBLib() { openvdb::uninitialize(); }
+    } vdblib_init;
+}
+
+template <typename PartioType, typename VDBType, typename IndexType> void
+convertScalar(const openvdb::points::AttributeHandle<VDBType>& vdbAttr,
+              const IndexType&                                 vdbIndex,
+              ParticlesDataMutable*                            particles,
+              ParticleAttribute&                               partioAttr,
+              size_t                                           partioIndex)
+{
+    auto* attrStorage = particles->dataWrite<PartioType>(partioAttr, partioIndex);
+    for (int i = 0; i < partioAttr.count; ++i)
+        attrStorage[i] = vdbAttr.get(vdbIndex, i);
+}
+
+template <typename PartioType, typename VDBType, typename IndexType> PartioType*
+convertVector(const openvdb::points::AttributeHandle<VDBType>& vdbAttr,
+              const IndexType&                                 vdbIndex,
+              ParticlesDataMutable*                            particles,
+              ParticleAttribute&                               partioAttr,
+              size_t                                           partioIndex)
+{
+    openvdb::Vec3d vdbData = vdbAttr.get(vdbIndex);
+    PartioType* partioData = particles->dataWrite<PartioType>(partioAttr, partioIndex);
+
+    for (unsigned i = 0; i < 3; ++i)
+        partioData[i] = vdbData[i];
+
+    return partioData;
+}
+
+enum {
+    Precision32,
+    Precision64
+};
+
+ParticlesDataMutable* readVDB(const char* filename,
+                              const bool headersOnly,
+                              std::ostream* errorStream)
+{
+
+    std::size_t numThreads = 0;
+    char *cueThreads = getenv("CUE_THREADS");
+    if (cueThreads)
+    {
+        numThreads = atoi(cueThreads);
+    }
+    tbb::task_scheduler_init schedulerInit((numThreads == 0) ? tbb::task_scheduler_init::automatic : numThreads);
+
+    initLibrary();
+
+    openvdb::io::File file(filename);
+    file.open();
+    if (!file.isOpen()) {
+        if (errorStream)
+            *errorStream << "Partio.VDB: Unable to open VDB file: '" << filename << "'\n";
+        return nullptr;
+    }
+
+    // Since we can be merging multiple VDB-Points objects:
+    // 1:
+    //   Walk the file collecting all the VDB-Point objects
+    //   Save the point attributes into an AttributeCache to ensure all attribute types match
+    //   Cache the index to the attribute (and it's precision)
+    // 2:
+    //   Re-walk all the found grids, copying data (using the cached attribute index and precision)
+    //
+    using AttributeIndex = std::tuple<size_t, ParticleAttributeType, bool, bool>;
+    struct AttributeCache {
+        std::vector<AttributeIndex> indices;
+        ParticleAttribute           partio;    // partio.type contains the type that partio will store / allocate
+        unsigned                    tupleSize;
```

Comment:
I dunno if it's just me, but I prefer seeing `unsigned int` over `unsigned` even though I know it's implied.